Add support for Adafruit QTPy RP2040 (and other boards)

smittytone opened the pull request:
I'm not offering this for merging per se, more to see if you're interested in pull in updates like this: to support other RP2040-based devices.

Happy to make changes too.

Cheers,

Tony

lurch commented on CMakeLists.txt:
```diff
@@ -6,13 +6,17 @@ project(picoprobe)
 
 pico_sdk_init()
 
+# Comment out to use Pico LED
+add_definitions(-DUSE_WS2812)
```
This should probably be something that's off-by-default that you then need to turn on for the Adafruit QTPy2040, rather than something that's on-by-default that you then need to turn off for the Raspberry Pi Pico? :wink:

smittytone replied:
That seems very reasonable

lurch commented on CMakeLists.txt:
```diff
@@ -6,13 +6,17 @@ project(picoprobe)
         src/probe.c
         src/cdc_uart.c
         src/get_serial.c
+        src/ws2812.c
```
Does it make sense to make the compilation of this (and also `ws2812.pio`) conditional on the value of `USE_WS2812` ? :shrug:

smittytone replied:
Will do.

lurch commented on src/picoprobe_config.h:
```diff
@@ -48,9 +48,19 @@
+#define PROBE_PIN_SWCLK PROBE_PIN_OFFSET + 0
+#define PROBE_PIN_SWDIO PROBE_PIN_OFFSET + 1
+
+#define NEO_SM 1
+#define NEO_PIN_PWR 11
+#define NEO_PIN_DAT 12
```
These should probably be in `CMakeLists.txt` rather than this header file, just like `PICOPROBE_LED` ?
See also https://github.com/raspberrypi/pico-sdk/blob/master/src/boards/include/boards/adafruit_qtpy_rp2040.h

smittytone replied:
Moved into `ws2812.h`.

lurch commented on src/picoprobe_config.h:
```diff
@@ -48,9 +48,19 @@
-#define PROBE_PIN_SWCLK PROBE_PIN_OFFSET + 0 // 2
-#define PROBE_PIN_SWDIO PROBE_PIN_OFFSET + 1 // 3
+#else
+#define PROBE_PIN_OFFSET 22 // Use the QTPy's QT STEMMA
```
Feels like this is kinda orthogonal to the `USE_WS2812` setting? (e.g. there might be some non-QTPy board that wants to use WS2812?)

Hmmm, I never noticed that https://learn.adafruit.com/adafruit-qt-py-2040/pinouts doesn't explicitly spell out which GPIO pins are used by the STEMMA connector - is that something that you could get added @dhalbert ?

dhalbert replied:
In the past we have supplied Arduino and CircuitPython pin names/numbers in our Learn Guides, and avoided supplying the "raw" chip pin numbers, because it tends to be confusing and increases support issues. The raw pin numbers are available in source, or more easily, on the schematic. GIven pico-sdk, though, we might reconsider this. I will ask @ladyada.

lurch replied:
Perhaps this would be less of an issue if https://github.com/raspberrypi/pico-sdk/blob/master/src/boards/include/boards/adafruit_qtpy_rp2040.h contained the STEMMA etc. pin numbers? :shrug:

smittytone replied:
FYI, Stemma pins grabbed from this: https://blog.adafruit.com/2021/05/20/pin-reference-adafruit-qt-py-rp2040/ which is the only place I could find it (see first comment above)

dhalbert replied:
For now, the schematics are available on the download pages in the Learn Guides for each board: https://learn.adafruit.com/adafruit-qt-py-2040/downloads.

The pin mappings in CircuitPython, which correspond to the silkscreen mappings, are in the `pins.c` file for each board, e.g. https://github.com/adafruit/circuitpython/blob/main/ports/raspberrypi/boards/adafruit_qtpy_rp2040/pins.c.

I'm not sure why we haven't added that colorful pin diagram to the QT Py 2040 guide yet. I think that may be just an oversight.

lurch commented on CMakeLists.txt:
```diff
@@ -16,11 +17,13 @@ add_executable(picoprobe
+        src/ws2812.c
 )
 
+# This stanza isn't working -- why not?
+# Plan is to include QTPy WS2812-specific Cmakle cmds here
```
I think your `add_definitions(-DUSE_QPTY_SWD` _only_ adds definitions visible to the compiler, whereas you need to use some other syntax if you want something that's visible/usable by cmake itself?

smittytone commented:
Updated with better device selection: the user sets the value of `USE_QTPY` in `CMakeLists.txt` and CMake config and then building, etc. proceed accordingly. Default is Picoprobe on Pico (`USE_QTPY` == `0`)

lurch commented on CMakeLists.txt:
```diff
@@ -1,43 +1,45 @@
-# Plan is to include QTPy WS2812-specific Cmakle cmds here
-if (DEFINED USE_QPTY_WS2812)
+if (USE_QTPY EQUAL 1)
+    message("Building for Adafruit QTPy RP20204 with WS2812")
```
RP20204 looks like a typo :wink:

lurch commented on CMakeLists.txt:
```diff
@@ -1,23 +1,40 @@
 
 pico_sdk_init()
 
+# Keep at 0 for Raspberry Pi Pico
+# Set this to 1 for QTPy with WS2812
+# Set this to 2 for QTPy without WS2812
+set(USE_QTPY 0)
```
Does `cmake` also allow you to set and compare strings? Having a `USE_QTPY` variable set to different numbers feels a bit arbitrary, and would obviously be hard to extend if somebody wanted to add support for a different RP2040 board?

smittytone replied:
Done

lurch replied:
IMHO that looks **much** better now :+1:

thezim commented:
I think using the default board [defines](https://github.com/raspberrypi/pico-sdk/tree/master/src/boards/include/boards) instead of a new define is a better route. On most of the existing boards define `PICO_DEFAULT_I2C_SDA_PIN` map to the STEMMA/QWIIC connector already and hard definitions in `picoprobe` are not needed. Defining PICO_BOARD sets it all in motion, see section `2.6.1. Preprocessor Variables via Board Configuration File` of the SDK [documentation](https://datasheets.raspberrypi.org/pico/raspberry-pi-pico-c-sdk.pdf).

Here is an example I tested.

```header
#ifdef ADAFRUIT_QTPY_RP2040 || SPARKFUN_PROMICRO
    #define PROBE_PIN_OFFSET PICO_DEFAULT_I2C_SDA_PIN
#else
    #define PROBE_PIN_OFFSET 2
#endif
```

However `PICO_DEFAULT_LED_PIN `becomes undefined. But on the plus sided `PICO_DEFAULT_WS2812_PIN` is defined for board that have them. Some boards like the SparkFun Thing Plus has both and in that case I say the the default LED should be used. I have not dug too deep but there may be boards with no available LED, regular to WS281x. The WS281x only is a case to like it is for my SparkFun Pro Micro.

kilograham commented:
yes, better to use the default pin defines so it works on everything
`PICO_DEFAULT_WS2812_PIN` and `PICO_DEFAULT_LED_PIN` should be defined on all boards that have them and undefined on those that don't.
Some of the examples to do with toggling LED pins for example will output a compiler warning on boards that don't have a LED pins

There is also a placeholder issue https://github.com/raspberrypi/pico-sdk/issues/188 to create a simple abstraction for signaling status/something on either LED or WS2812

smittytone commented:
As per above, I've updated config to use `PICO_BOARD` as the initialisation point, with the standard value entered there to define what's needed for a specific board. I've added a couple more — @dhalbert’s SparkFun Pro Micro and the Adafruit Feather, the focus being on boards I have with STEMMA ports — and made use of SDK-set #define values wherever possible. One case were it's not: the GPIO pin number for SWD CLK on the QTPy which is not the same as the standard SDA pin.

I've also brought the RGB LED colour setting into `CMakeLists.txt` to reduce the number of edits a user might have to make to other files if they're building for a known board.

Added a Read Me too, as that's what I do for a living.

PS. I've crudely 'fixed' the undefined `PICO_DEFAULT_LED_PIN` issue, but there must be better approach that doesn't send folk to the command line, surely?

lurch commented on CMakeLists.txt:
```diff
@@ -1,31 +1,64 @@
+endif()
+
+if (NOT PICO_BOARD STREQUAL "pico")
+    target_sources(${EXECUTABLE_NAME} PRIVATE src/ws2812.c)
```
This seems to be assuming that any board which **isn't** a Raspberry Pi Pico will _always_ have a WS2812 LED, which I guess won't always be the case? Probably better to do something like `set(HAS_WS2812 1)` in the bit above where you're checking individual board-names?

lurch commented on README.md:
```diff
@@ -0,0 +1,25 @@
+# Picoprobe
+
+A Picoprobe is a Raspberry Pi RP2040-based Single-Wire Debug (SWB) adaptor. You can make one from a Raspberry Pi Pico — hence the name — or almost any third-party RP2040-based board. Use it to connect your RP2040-based application board to your personal computer for debugging through `gdb`, directly or via an IDE such as Microsoft Visual Studio Code.
```
Single-Wire Debug -> Serial Wire Debug
SWB -> SWD

lurch commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+A Picoprobe is a Raspberry Pi RP2040-based Single-Wire Debug (SWB) adaptor. You can make one from a Raspberry Pi Pico — hence the name — or almost any third-party RP2040-based board. Use it to connect your RP2040-based application board to your personal computer for debugging through `gdb`, directly or via an IDE such as Microsoft Visual Studio Code.
+
+The original version of this code only supported the Pico. This update adds support for three more boards, each of which provides a [STEMMA QT connector](https://learn.adafruit.com/introducing-adafruit-stemma-qt/what-is-stemma-qt)/[Quicc](https://www.sparkfun.com/qwiic) for easy cabling to the application board under test. These connectors are intended for use with I&sup2;C devices, but here we’r ...
```
I don't think it makes sense for README.md to talk about "This update..." ? (this is the README, not a CHANGELOG)
Quicc -> Qwiic ?

thezim replied:
I think what is worth mentioning is the that if you are using the STEMMA/Qwiic board is that UART lines will not be passed through. Using the base board has that advantage as you don't need to plug in both your pico probe and the target board to USB.

lurch commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+The original version of this code only supported the Pico. This update adds support for three more boards, each of which provides a [STEMMA QT connector](https://learn.adafruit.com/introducing-adafruit-stemma-qt/what-is-stemma-qt)/[Quicc](https://www.sparkfun.com/qwiic) for easy cabling to the application board under test. These connectors are intended for use with I&sup2;C devices, but here we’r ...
+
+The code also uses the WS2812 RGB LED on those boards that have one as an SWD activity indicator.
```
Perhaps "on those boards that have one" should be in parentheses?

lurch commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+The code also uses the WS2812 RGB LED on those boards that have one as an SWD activity indicator.
+
+If your chosen board is not one of those so far supported by the Picoprobe code, you should be able to use with just a few small code changes.
```
use with -> use it with ?

lurch commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+## Build the Picoprobe code
+
+To build for your own Picoprobe:
```
To build Picoprobe for your own board:" ?

lurch commented on CMakeLists.txt:
```diff
@@ -1,31 +1,64 @@
+    message("Building for Adafruit Feather RP2040 with WS2812")
+elseif (PICO_BOARD STREQUAL "sparkfun_promicro")
+    message("Building for SparkFun Pro Micro RP2040 with WS2812")
+else()
```
Should this `else()` be `elseif (PICO_BOARD STREQUAL "pico")` ?

lurch commented on src/picoprobe_config.h:
```diff
@@ -65,9 +75,11 @@
 #ifndef PICOPROBE_LED
 
 #ifndef PICO_DEFAULT_LED_PIN
-#error PICO_DEFAULT_LED_PIN is not defined, run PICOPROBE_LED=<led_pin> cmake
+// #error PICO_DEFAULT_LED_PIN is not defined, run PICOPROBE_LED=<led_pin> cmake
+#define PICOPROBE_LED 0     // To avoid compiler errors -- is there a better way to do this?
```
Perhaps you want something like `#if !defined(PICO_DEFAULT_LED_PIN) && !defined(PICO_DEFAULT_WS2812_PIN)` ?

thezim commented on CMakeLists.txt:
```diff
@@ -1,31 +1,64 @@
+#    "adafruit_feather_rp2040" for the Adafruit Feather RP2040
+#    "sparkfun_promicro"       for the SparkFun Pro Micro RP2040
+
+set(PICO_BOARD "pico")
```
Should this be defined in this file? A PICO_BOARD environment variable works too. Not having to modify any files make building even easier.

Linux/MacOS

```bash
export PICO_BOARD=SPARKFUN_PROMICRO
cmake .
```

WIndows
```batch
SET PICO_BOARD=SPARKFUN_PROMICRO
cmake .
```

PowerShell Core (All OS'es)
```powershell
$env:PICO_BOARD=SPARKFUN_PROMICRO
cmake .
```

thezim commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+To build for your own Picoprobe:
+
+1. Edit the `CmakeLists.txt` file and change the value of  `PICO_BOARD` in line 16.
```
Line number change over time. Better to reference this by name and not by line number. Less work later.

smittytone commented:
I've closed it because I don't think it's getting anywhere. I opened the request to see if there was any interest on the part of the Foundation in extending this. That question hasn't been answered. I've had some useful suggestions for improvements, which I have applied, and a fair few opinions, but nothing in the way of direction, and I think that's what is required at this stage. Someone with the authority to do so needs to set a framework for how other devices get added, and oversee their addition as punters submit them. I proposed one approach, and anyone can view my fork to see how it works. I'm open to others, and I'm also happy to reopen this PR if the Foundation thinks that's the way to go.

lurch commented:
ping @kilograham and @liamfraser
